### ISSUE TITLE
thisProcess.newExecutingPath is not updated immediately from the IDE

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#4607]
## Environment

* SuperCollider version: 3.10.3
* Operating system: macOS 10.13
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce
This shows the cause of the problem:
```supercollider
// nowExecutingPath test
// 1. save this as a .scd document
// 2. run these lines separately in short succession
thisProcess.recompile;
thisProcess.nowExecutingPath.postln; //evaluate this line before recompile finishes
// -> nil

// 3. evaluate this again after the recompile is done
thisProcess.nowExecutingPath.postln; //evaluate this line before recompile finishes
// -> proper path
```
The problem can be clearly seen if the code that's run right after recompiling depends on `thisProcess.nowExecutingPath`:

```supercollider
// resolveRelative test
// 1. save this as a .scd document
// 2. run these lines separately in short succession
thisProcess.recompile;
"".resolveRelative; //evaluate this line before recompile finishes
// -> ERROR
```

## Expected vs. actual behavior

`thisProcess.nowExecutingPath` and thus `"a/path".resolveRelative` should be available to the user right after recompiling the library (or before any other code is executed from the IDE). Currently it takes a short time before that value is available. 

The case where this is problematic is when the code includes `"".resolveRelative` and the user wants to re-start the system, hitting the keyboard shortcut to recompile the library and right away attempting to run the code. This result is `ERROR: can't resolve relative to an unsaved file` and the execution stops.

**Comment by brianlheim** [comment]
i don't believe this has anything to do with SCIDE, does it?

**Comment by dyfer** [comment]
> i don't believe this has anything to do with SCIDE, does it?

My understanding is that `thisProcess.nowExecutingPath` is updated from the IDE - the interpreter is not running a file but rather it is used interactively, isn't it?

**Comment by jamshark70** [comment]
https://github.com/supercollider/supercollider/blob/9f6bc367ba7d42002e7b3c5261e824ec3031ed07/SCClassLibrary/Common/Core/Kernel.sc#L613

```
thisProcess.nowExecutingPath = ideClass.currentPath
```

I believe the IDE issues `ScIDE.currentPath = "..."` commands directly.

In ScIDE *handshake, there's 
`this.send(\requestCurrentPath);` -- this is asynchronous so of course there is no way to make it absolutely immediate.

What we would have to do is block `interpretPrintCmdLine` until all initialization is complete. We can't stop the user from submitting commands, but we can queue them until the environment is fully ready to run them.

**Comment by jamshark70** [comment]
Btw one *could* argue it's an invalid use case to run code while the library is still compiling :wink: (though one could argue back that the current handling of that case, effectively to fail silently, is at best ungraceful).

*But* we have just fixed document autoRun, which will certainly be affected by the issue. It's definitely bad to tell users they can run init at recompile time as a supported feature, but relative paths will be unavailable at that time.

**Comment by brianlheim** [comment]
ok, right, this is abstracted over environments, so I will remove the tag.

blocking execution until the library is fully recompiled seems not only a good solution, but how it should have been from the beginning

**Comment by dyfer** [comment]
> Btw one _could_ argue it's an invalid use case to run code while the library is still compiling 😉 

I think it would be acceptable if the code wouldn't run at all, but it is effectively "queued" and run once the recompile is done (but not `thisProcess.nowExecutingPath` is not ready).

>blocking execution until the library is fully recompiled seems not only a good solution, but how it should have been from the beginning

Agreed.

**Comment by dyfer** [comment]
@jamshark70 @brianlheim thinking towards a solution, which behavior seems more reasonable?
When the user attempts running the code while the interpreter is recompiling the library:
1. nothing happens; user needs to run the code again once the interpreter is ready
2. the code is run immediately after the recompile is done (current behavior)

**Comment by jamshark70** [comment]
> Which behavior seems more reasonable, when the user attempts running the code while the interpreter is recompiling the library?

Ideal solution is to run the code immediately after the environment is ready (after all initialization is finished) -- meaning we need some signaling (which we never did before). Maybe a global Condition and global counter, and all asynchronous init (including in quarks) should register with it and unregister when finished. That is:

1. Set global counter for pending init tasks and `Condition({ counter <= 0 })`.
2. Run initClassTree.
3. Within initClassTree, any asynchronous process should increment the counter, and decrement when finished. (Timeout/error handling become important, then.)
4. interpretPrintCmdLine queues up the command and does `condition.wait` (which doesn't wait, if the condition is satisfied).

If we can't do that, then we should print a warning and ask the user to do it again.

Worst case is the current behavior (IMO).

**Comment by dyfer** [comment]
@jamshark70 thanks for the detailed answer. However if you use the condition, wouldn't we need to run that whole thing in a `Routine`? This would result in a drastic change to the behavior of running code in the described case (code would be forked), vs running normally (code is not forked).

**Comment by jamshark70** [comment]
> However if we use the condition, wouldn't we need to run that whole thing in a Routine? This would result in a drastic change to the behavior of running code in the described case (code would be forked), vs running normally (code is not forked).

The other way would be to implement the blocking either in sclang (don't call interpretPrintCmdLine at all until init is finished, and handle the queuing in c++) or in the IDE (sclang should send a signal when all init is finished, and the IDE would submit commands only after that).

Of those two, the first is probably better. (We should improve signaling to the IDE in any case; handshaking is rather poor presently, with a number of issues open now describing race conditions.)

In my ~~proposal~~ very loose, off the cuff suggestion, we could check the condition and, if true, interpret the command immediately outside of a routine. (It would be risky to move all evaluation into a routine, but... what, concretely, would break? "Drastically" is a large assertion.) That sort of check would eliminate the main concern 

That leaves commands submitted in the few second window during library compilation. It's hard to imagine a serious consequence if these run in an AppClock routine.

**Comment by dyfer** [comment]
> "Drastically" is a large assertion.) That sort of check would eliminate the main concern.
> 
> That leaves commands submitted in the few second window during library compilation. It's hard to imagine a serious consequence if these run in an AppClock routine.

Yeah, maybe "Drastically" is an exaggeration. However I have experienced many subtle bugs (in my code, not in SC) when there was additional forking which made the commands run in a different order than anticipated. But maybe, indeed, it's not the end of the world.

However, this would possibly create a more visible change for autorun code using `/*RUN*/`, as it would always be run in a routine when recompiling library, but not when opening the document (I think).

**Comment by jamshark70** [comment]
> However I have experienced many subtle bugs (in my code, not in SC) when there was additional forking which made the commands run in a different order than anticipated.

That's true -- but I had suggested that incoming commands during this window should be queued (and not simply fork a new routine on every call). First in first out would handle the order.

> However, this would possibly create a more visible change for autorun code using /*RUN*/, as it would always be run in a routine when recompiling library, but not when opening the document (I think).

Valid point. Would it be valid to fork autorun when opening the document too? That might be the best way.

**Comment by totalgee** [comment]
I've also run into this problem quite often recently, and found it annoying. But thanks to the thread above, at least I see that it's "solved" if I'm a bit slower on hitting Ctrl-Enter after recompiling the class library. The misleading thing is that the "Welcome to SuperCollider" message appears before the "can't load relative to an unsaved file" error, so I assumed that sclang was "good to go"...

```
*** Welcome to SuperCollider 3.10.3. *** For help press Ctrl-D.
ERROR: can't load relative to an unsaved file
CALL STACK:
	Exception:reportError
		arg this = <instance of Error>
	Nil:handleError
		arg this = nil
		arg error = <instance of Error>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of Error>
	Object:throw
		arg this = <instance of Error>
	String:loadRelative
		arg this = "../helpers/liveCodingSetup.scd"
		arg warn = true
		arg action = nil
		var path = nil
	< closed FunctionDef >
		var postBootFunc = <instance of Function>
		var liveCodingSetup = nil
	Interpreter:interpretPrintCmdLine
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "(
//HelpBrowser.openHelpFor(..."
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine
		arg this = <instance of Main>
^^ The preceding error dump is for ERROR: can't load relative to an unsaved file
```

Thanks, I'm glad to see this is logged.

**Comment by prko** [comment]
@dyfer I think this issue can be closed as I have tested this on all platforms, sc returns the correct path immediately after recompling. How about on your machine?

**Comment by dyfer** [comment]
@prko @telephon can you provide details on the SC version you run this under?

```
// run these lines separately in short succession, but NOT together
thisProcess.recompile;
// evaluate this line before recompile finishes, i.e. WHILE recompile is happening
thisProcess.nowExecutingPath.postln;
```

The following still produces `nil` for me for both SC 3.13 as well as a recent develop build under macOS 14.
```
(...)
*** Welcome to SuperCollider 3.14.0-dev. *** For help press Cmd-D.
nil
-> nil
```

**Comment by prko** [comment]
My SC is built from the 'develop' branch [d52b052], and I am using it on MacOS 14.7.5.

The evaluation of your code snippet returns `nil` as you described.

**Comment by telephon** [comment]
Same, with 15.3.2 .

**Comment by telephon** [comment]
**Wait**, now I see that I may have misunderstood the descropition above.
I have added the following to my startup file:

```supercollider
"In startup file: thisProcess.nowExecutingPath: %".format(thisProcess.nowExecutingPath).postln;
StartUp.add({ "In StartUp class: thisProcess.nowExecutingPath.: %".format(thisProcess.nowExecutingPath).postln; });
```

And indeed it posts:
```supercollider
In startup file: thisProcess.nowExecutingPath: /Users/wanninger/Library/Application Support/SuperCollider/startup.scd
In StartUp class: thisProcess.nowExecutingPath.: nil
```

This is wrong ... but also interesting. The call in the startup file comes *before*, and returns the right path. The one in the StartUp class is later, but `nil`. 

When I then call it by hand:

```supercollider
StartUp.run;
```

Then it returns the path I am in at that moment.

**Comment by telephon** [comment]
The best reproducer so far is:
```supercollider
"In startup file: thisProcess.nowExecutingPath: %".format(thisProcess.nowExecutingPath).postln;
AppClock.sched(0.01, { "In startup file: thisProcess.nowExecutingPath: %".format(thisProcess.nowExecutingPath).postln; });
```

Returns: on startup
```
----------- end all server processes -------------
In startup file: thisProcess.nowExecutingPath: /Users/j/Library/Application Support/SuperCollider/startup.scd


*** Welcome to SuperCollider 3.14.0-dev. *** For help press Cmd-D.
In startup file: thisProcess.nowExecutingPath: nil
```

**Comment by telephon** [comment]
OK, part of the riddle is solved:

The path exists in the startup file, because it is set in `path.load`, which calls `this.executeFile`. It only ever exists where something is called through interpret or compile.

**Comment by dyfer** [comment]
> The best reproducer so far is, in the startup file:

Sure, that works.

FWIW I experience this most when restarting system for a clean run before a soundcheck/performance etc. The workflow is
1. recompile the library (cmd-shift-L)
2. run all code (cmd-enter)
then way for the system to start

If I don't wait before step 2, I get errors if my code depends on knowing its path.

**Comment by jamshark70** [comment]
As I understand it, the boundary conditions for nowExecutingPath are:

- Interactive code execution: interpretPrintCmdLine sets to the current Document path.
- `.load`, `executeFile`: Sets to the file path.
- Scheduled Routines: The path is tracked in the Routine object.
- Scheduled or "executed-later" functions **are not supported for nowExecutingPath** and currently cannot be.

So the startup.scd vs StartUp mystery is simply that normal StartUp use is to add a function, which will be executed later at the right time. Functions do not remember the path in which they were created. I've seen multiple versions of the same issue being raised in this thread, and every time it's raised, it's "a new issue" and "weird behavior" -- but they all stem back to the use of functions.

To fix this would require adding a `path` or `filenameSymbol` member variable to FunctionDef (I guess...?) -- perhaps move it out of Method -- which would be populated at compile time.

So, do we want to do that surgery?

> ```
> // run these lines separately in short succession, but NOT together
> thisProcess.recompile;
> // evaluate this line before recompile finishes, i.e. WHILE recompile is happening
> thisProcess.nowExecutingPath.postln;
> ```

I would not expect all language features to work if statements are submitted before initialization has completed.

Perhaps the fix here is for sclang to disallow interactive code execution until the entire startup sequence has finished.

**Comment by dyfer** [comment]
> I would not expect all language features to work if statements are submitted before initialization has completed.

Sure, but SC states that it _is_ ready (`Welcome to SuperCollider...`), when it isn't, not really. And running the code during recompile is not blocked, but only delayed... it should either fully work IMO, or execution should be blocked until after the recompile, as you suggested (I would prefer the former though).

And thanks for the reminder on `nowExecutingPath` being unsupported insider Routines etc. I'm well aware of it, yet I didn't catch that it was the issue in Julian's example.

BTW the reminder on name resolution inside routines also raises a point about possibly running class init in a routine (at least _I_ thought about doing that at some point); that would break the use of `nowExecutingPath` inside the Class init, which is a no-go, I'd think.

**Comment by telephon** [comment]
There is a simple example that shows that @jamshark70 is correct in his diagnosis. Wait for supercollider to start up completely. Then paste this into a document that has been saved and run it:

```supercollider
(
"directly called, path is: %".format(thisProcess.nowExecutingPath).postln;
Routine { 1.wait; "scheduled in routine, path is: %".format(thisProcess.nowExecutingPath).postln }.play(AppClock);
AppClock.sched(0.5, { "scheduled, path is: %".format(thisProcess.nowExecutingPath).postln });
)
```
The first two return a path, the third doesn't.

I don't think that the issue has anything to do with the IDE or the startup process. But maybe you have a different diagnosis.
In any case, I've made another issue #6675 that focusses on that problem specifically.

**Comment by muellmusik** [comment]
> To fix this would require adding a `path` or `filenameSymbol` member variable to FunctionDef (I guess...?) -- perhaps move it out of Method -- which would be populated at compile time.

IIUC I can see a potential confusion with adding it to FunctionDef. Eg:

1. User defines a function in a saved file. 
2. User later evaluates that function in a different file. 

If we refer to a path stored in the function def we get file 1, but it would be reasonable for the user to expect it to be file 2.

Perhaps the semantics could be clarified, or there's a more elegant solution, but maybe an acceptable option is to document that if you need this for a deferred function capture it like so:

```
var path = thisProcess.nowExecutingPath;

~myFunc = { do something with path };
```

**Comment by telephon** [comment]
> IUC I can see a potential confusion with adding it to FunctionDef. Eg:
> 
>     1. User defines a function in a saved file.
> 
>     2. User later evaluates that function in a different file.

Yes, I was about to write something about the semantics of "now executing". It is similar to the difference between lexical and dynamic scope.

To make it clear, should we parhaps define two variables – `sourcePath` (or `filenameSymbol`, like in Method) and `nowExecutingPath`. Resolving a path with loadRelative should use the `sourcePath` and not the `nowExecutingPath`.

**Comment by jamshark70** [comment]
> Sure, but SC states that it is ready (Welcome to SuperCollider...), when it isn't, not really. And running the code during recompile is not blocked, but only delayed... it should either fully work IMO, or execution should be blocked until after the recompile, as you suggested (I would prefer the former though).

I can't think of any reason to allow the interpreter to try to execute interactive code before it has finished starting up. That just seems to open the door to all kinds of risks.

I would be OK with putting interactive code into a queue, to be evaluated immediately after the startup process is completely finished.

> And thanks for the reminder on nowExecutingPath being unsupported insider Routines etc

It *is* supported in Routines -- not Functions.

> IIUC I can see a potential confusion with adding it to FunctionDef. Eg:
> 
> 1. User defines a function in a saved file.
> 2. User later evaluates that function in a different file.
>
> If we refer to a path stored in the function def we get file 1, but it would be reasonable for the user to expect it to be file 2.

The trouble here is that sometimes a user will expect 1, and other times 2.

For myself, I basically never want 2 -- only 1. If I'm defining some functionality and it needs to refer to a relative path, I want it to be relative to the definition, not relative to some arbitrary place from which it was called. I realize I'm biased toward my own use cases, but option 2 seems to me to invite chaos. The file defining the functions/routines has no control over the places from which those will be called -- in case 2, then, pretty much everything related to nowExecutingPath would be up for grabs. I personally will take great care never ever to rely on that.

Perhaps nowExecutingPath is simply misnamed, then.

**Comment by telephon** [comment]
> Perhaps nowExecutingPath is simply misnamed, then.

Both cases are justified, I think: you may want to know where a function was defined and also you may want to know where it was called.

But I see this insight as a way forward – keeping a separate variable for the former.

**Comment by JordanHendersonMusic** [comment]
Regards an implementation for moving the file path to functiondef. 

I thought I'd have a go at outlining an approach, as I've been caught out by this confusion.

It seems ides communicate with the VM through a separate process, meaning the current active file path needs to be sent from the ide *in SC code*.

https://github.com/supercollider/supercollider/blob/develop/editors%2Fsc-ide%2Fcore%2Fsc_process.cpp#L212

I'm a bit unsure how to do this, as ideally, you'd want the VM to allow the current compiling path to be set outside of the language (this way the parser would know what file is active).

Here is an option. 

All code to be evaluated from the ide could be preceded by a silent command (meaning doesn't post)
```
FunctionDef.pushCurrentCompilingPath(...);
```
and succeeded by 
```
FunctionDef.popCurrentCompilingPath();
```

This is isn't ideal, as the above code's current compiling path will be nil. This is a catch 22 situation that I don't you can avoid, but it also isn't user code so not that big of a deal.

Then the compiler can look at the stack of paths in a class var of functiondef and get the last path.

Now a question remains over what `Strign.load` should do? Is that loading a new file, or injecting that code into the current file. I can see a use case for both approaches.
In the former case, it should push and pop a new path.

**Comment by prko** [comment]
> Perhaps the semantics could be clarified, or there's a more elegant solution, but maybe an acceptable option is to document that if you need this for a deferred function capture it like so:
> 
> ```
> var path = thisProcess.nowExecutingPath;
> 
> ~myFunc = { do something with path };
> ```

@muellmusik 
There is a PR for this: #6291 

The following file is a rendered PDF of Process.schelp:
[Process _ SuperCollider 3.14.0-dev Help.pdf](https://github.com/user-attachments/files/19402274/Process._.SuperCollider.3.14.0-dev.Help.pdf)

However, even after understanding the concept, it is still sometimes confusing when using ...

**Comment by muellmusik** [comment]
I think @telephon’s suggestion here is probably the best way forward. The semantics of `nowExecutingPath` _should_ be clear, and I can see both use cases. So adding `sourcePath` and a bit of documentation should clear things up I hope?

**Comment by dyfer** [comment]
> I can't think of any reason to allow the interpreter to try to execute interactive code before it has finished starting up. That just seems to open the door to all kinds of risks. The current behavior strikes me as a bug.

@jamshark70 I'd argue that the interpreter is not trying to execute the code before it has finished starting up - it executes it right after starting up. The problem is that the path is not updated until slightly _after_ the interpreter is started, IIUC.

> It _is_ supported in Routines -- not Functions.

Ah, thank you for clarifying! 

-----

Everyone, this issue is about newExecutingPath not being populated until after sclang claims to be ready. Let's move the discussion on clarifying newExecutingPath and possibly adding path functionality to a FunctionDef to a separate issue. Thanks!

**Comment by telephon** [comment]
> Everyone, this issue is about `nowExecutingPath` not being populated until after sclang claims to be ready. Let's move the discussion on clarifying `nowExecutingPath` and possibly adding path functionality to a FunctionDef to a separate issue. Thanks!

This is why I've made a separate issue and linked it, here it is again: https://github.com/supercollider/supercollider/issues/6675

**Comment by JordanHendersonMusic** [comment]
Sorry I missed that!
